source/tutorials/cross-compilation.rst

```diff
@@ -0,0 +1,257 @@
+    helloWorld = pkgs.writeText "hello.c" ''
+      #include <stdio.h>
+      
+      int main (void)
+      {
+        printf ("Hello, world!\n");
+        return 0;
+      }
+    '';
+
+    # A function that takes host platform packages
+    crossCompileFor = hostPkgs:
+      # Run a simple command with the compiler available
+      hostPkgs.runCommandCC "hello-world-cross-test" {} ''
+        # Wine requires home directory
+        HOME=$PWD
+
+        # Compile our example using the compiler specific to our host platform
+        $CC ${helloWorld} -o hello 
+
+        # Run the compiled program using an emulator
+        # Usually Qemu, but on windows it is Wine
+        ${hostPkgs.stdenv.hostPlatform.emulator hostPkgs.buildPackages} hello > $out
+
+        # print to stdout program stdout
+        cat $out
+      '';
+  in {
+    # Statically compile our example using the two platform hosts
+    rpi = crossCompileFor pkgs.pkgsCross.raspberryPi;
+    windows = crossCompileFor pkgs.pkgsCross.mingwW64;
+  }
+
+If we build this example and print both resulting derivations, we should see "Hello, world!" for each: 
+
+.. code:: shell-session
+
+  $ cat $(nix-build cross-compile.nix)
+  Hello, world!
+  Hello, world!
+
+
+Developer environment with a cross-compiler
+-------------------------------------------
+
+In :ref:`tutorial for declarative reproducible environments <declarative-reproducible-envs>`,
+we've looked at how Nix helps us provide tooling and system libraries for our project.
+
+It's also possible to provide an environment with a compiler configured for cross-compilation!
+
+Given we have a ``shell.nix``:
+
+.. code:: nix
+
+  { nixpkgs ? fetchTarball "https://github.com/NixOS/nixpkgs/archive/bba3474a5798b5a3a87e10102d1a55f19ec3fca5.tar.gz"
+  , pkgs ? (import nixpkgs {}).pkgsCross.aarch64-multiplatform
+  }:
+
+  # pkgs.callPackage is needed due to https://github.com/NixOS/nixpkgs/pull/126844
+  pkgs.callPackage ({ mkShell, zlib, pkg-config }: mkShell {
+    # these tools run on the build platform, but are configure to target the target platform
+    nativeBuildInputs = [ pkg-config file ];
+    # libraries needed for the target platform
+    buildInputs = [ zlib ];
+  }) {}
+
+And ``hello.c``:
+
+.. code:: c 
+
+  #include <stdio.h>
+
+  int main (void)
+  {
+    printf ("Hello, world!\n");
+    return 0;
+  }
+
+We can cross-compile it:
+
+.. code:: shell-session 
+
+  $ nix-shell --run '$CC hello.c -o hello' cross-compile-shell.nix
+
+And confirm it's aarch64:
+
+.. code:: shell-session 
+
+  $ nix-shell --run 'file hello' cross-compile-shell.nix 
+  hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /nix/store/733hzlw1hixdm6dfdsb8dlwa2h8fl5qi-glibc-2.31-74-aarch64-unknown-linux-gnu/lib/ld-linux-aarch64.so.1, for GNU/Linux 2.6.32, with debug_info, not stripped
+  
+
+Next steps
+----------
+
+- `Official binary cache <https://cache.nixos.org>`_ doesn't come with binaries
```

Comment:
This is technically not true, we have some cache for certain things as part of the `cross-trunk` jobset although it should probably be expanded one of these days…

Reply:
I wanted to avoid that to set expectations low and simplify explanation. Maybe something like `comes with almost no binaries`?

Reply:
How about something like “We have extremely limited binary-cache for cross compilation, so it is very possible that you have to compile the respective toolchain yourself”? This would also clarify the risk (compiling binutils and gcc).